### PR TITLE
GH-1082: Dead Letter Streams Deserialization

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-kafka/issues/1082

Needs docs.

### CODE REVIEW

File: spring-kafka/src/test/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandlerTests.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+			props.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, this.brokerAddresses);
+			props.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.ByteArraySerde.class);
+			props.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, FailSerde.class);
+			props.put(StreamsConfig.DEFAULT_TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
+					WallclockTimestampExtractor.class.getName());
+			props.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, "100");
+			props.put(StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG,
+					RecoveringDeserializationExceptionHandler.class);
+			props.put(RecoveringDeserializationExceptionHandler.KSTREAM_DESERIALIZATION_CONTEXT_ID, context.getId());
+			props.put(RecoveringDeserializationExceptionHandler.KSTREAM_DESERIALIZATION_RECOVERER, "recoverer");
```
Comment (artembilan) [inline spring-kafka/src/test/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandlerTests.java]: And why we can't just store a `recoverer()` instance in the properties to configure streams?

File: spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	public static final String KSTREAM_DESERIALIZATION_CONTEXT_ID = "spring.deserialization.context.id";
+
+	/**
+	 * Property name for the bean name of the recoverer.
+	 */
+	public static final String KSTREAM_DESERIALIZATION_RECOVERER = "spring.deserialization.recoverer";
+
+	private static final Log LOGGER = LogFactory.getLog(RecoveringDeserializationExceptionHandler.class);
+
+	private ConsumerRecordRecoverer recoverer;
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java]: `final` and `@Nullable` and also initialized as a `null` in the ctor.

File: spring-kafka/src/test/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandlerTests.java (hunk 1 above)
Comment (garyrussell) [reply spring-kafka/src/test/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandlerTests.java]: It doesn't work - see my comment in the issue.

File: spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java (hunk 2 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java]: ? It's assigned in `configure()`.

Comment (artembilan) [reply spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java]: Ah! That's all about `toString()` on objects in that case.

And there is no way to pass object at all:
```
public DeserializationExceptionHandler defaultDeserializationExceptionHandler() {
        return getConfiguredInstance(DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG, DeserializationExceptionHandler.class);
    }
```

Then OK: let's live with what you have so far. Until Apache Kafka will come up with the setters on the `KafkaStreams`...

Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/streams/RecoveringDeserializationExceptionHandler.java]: Actually, the problem is ours - `KafkaStreamsConfiguration.asProperties()` - we are doing conversion there. However, they are using the `Properties` as a general Map, with any object type. It's a bit of an abuse but I think we can change that method to just `putAll()`.

File: spring-kafka/src/main/java/org/springframework/kafka/config/StreamsBuilderFactoryBean.java [hunk 3]
```diff
@@ -66,6 +71,8 @@
 
 	private static final String CLEANUP_CONFIG_MUST_NOT_BE_NULL = "'cleanupConfig' must not be null";
 
+	private static final Map<String, ApplicationContext> contexts = new ConcurrentHashMap<>();
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/StreamsBuilderFactoryBean.java]: Look like we don't this logic any more.

File: spring-kafka/src/main/java/org/springframework/kafka/config/StreamsBuilderFactoryBean.java [hunk 4]
```diff
@@ -52,7 +53,8 @@
  *
  * @since 1.1.4
  */
-public class StreamsBuilderFactoryBean extends AbstractFactoryBean<StreamsBuilder> implements SmartLifecycle {
+public class StreamsBuilderFactoryBean extends AbstractFactoryBean<StreamsBuilder>
+		implements SmartLifecycle, ApplicationContextAware {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/config/StreamsBuilderFactoryBean.java]: This is not an `ApplicationContextAware` any more...

**Pull request merged**